File: python/Server/bookinfo_server.py
import asyncio
import logging
import json

import grpc
from bookinfo_pb2 import BookItem
from bookinfo_pb2 import SearchResponse
from bookinfo_pb2 import SearchRequest
from bookinfo_pb2_grpc import MultiBookInfoServicer
from bookinfo_pb2_grpc import add_MultiBookInfoServicer_to_server

#NUMBER_OF_REPLY = 10

USER_DB = "../Data/bookinfo.json"
with open(USER_DB, mode="r") as fp:
    books = json.load(fp)

class BookInfo(MultiBookInfoServicer):

    async def Search(self, request: SearchRequest,
                       context: grpc.aio.ServicerContext) -> SearchResponse:
        logging.info("Serving Search request %s", request)
        text = request.text
        for i in books:
            title = i["title"]
            author = i["author"]
            publisher = i["publisher"]
            if text in title or text in author or text in publisher:
                bookitem = BookItem(title=i["title"], author=i["author"],
                    publisher=i["publisher"], isbn=i["isbn"], price=i["price"],
                    type=i["type"])
                yield SearchResponse(status_code = 1, item = bookitem)

async def serve() -> None:
    server = grpc.aio.server()
    add_MultiBookInfoServicer_to_server(BookInfo(), server)
    listen_addr = "[::]:50051"
    server.add_insecure_port(listen_addr)
    logging.info("Starting server on %s", listen_addr)
    await server.start()
    await server.wait_for_termination()

if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(serve())

File: nodejs/Server/bookinfo_server.js
const fs = require('fs');
const _ = require('lodash');

const PROTO_PATH = __dirname + '/../Protos/bookinfo.proto';
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
const packageDefinition = protoLoader.loadSync(
    PROTO_PATH,
    {keepCase: true,
     longs: String,
     enums: String,
     defaults: true,
     oneofs: true
    });
const bookinfo = grpc.loadPackageDefinition(packageDefinition).bookinfo;

let bookinfo_list = [];

function search(call) {
  const keyword = call.request.text;
  _.each(bookinfo_list, (bookinfo) => {
    const title = bookinfo.title;
    const author = bookinfo.author;
    const publisher = bookinfo.publisher;
    if (title.includes(keyword) || author.includes(keyword) || publisher.includes(keyword)) {
      const item = {
        title: title, author: author, publisher: publisher,
        isbn: bookinfo.isbn, price: bookinfo.price, type: bookinfo.type
      };
      const res = {
        status_code: 1,
        item: item
      }
      call.write(res);
    }
  });
  call.end();
}

function register(call, callback) {
  let count = 0;
  call.on('data', (request) => {
    count += 1;
    bookinfo_list.push(request.item);
  });
  call.on('end', () => {
    callback(null, {
      count: count
    });
  });
}

function getServer() {
  const server = new grpc.Server();
  server.addService(bookinfo.BookInfo.service, {
    search: search,
    register: register
  });
  return server;
}

if (require.main === module) {
  const server = getServer();
  server.bindAsync('0.0.0.0:50051', grpc.ServerCredentials.createInsecure(), () => {
    const DB_PATH = __dirname + '/../Data/bookinfo.json';
    fs.readFile(DB_PATH, (err, data) => {
      if (err) throw err;
      bookinfo_list = JSON.parse(data);
      server.start();
    });
  });
}

exports.getServer = getServer;
